drivers: pwm_led: esp32: add inverted pwm polarity

epc-ake opened the pull request:
This PR enables the PWM to be inverted by making use of the `PWM_POLARITY_INVERTED` flag.
It also fixes an overflow bug when setting duty-cycle to 100 percent.

epc-ake committed:
drivers: pwm_led: esp32: add inverted pwm polarity

This PR enables the PWM to be inverted by making use of the
`PWM_POLARITY_INVERTED` flag.
It also fixes an overflow bug when setting duty-cycle to 100 percent.

Signed-off-by: Armin Kessler <ake@espros.com>

wmrsouza commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -279,6 +279,10 @@ static int pwm_led_esp32_set_cycles(const struct device *dev, uint32_t channel_i
 	}
 
+	if (flags & PWM_POLARITY_INVERTED) {
+		pulse_cycles = period_cycles - pulse_cycles;
```
Would you mind checking period_cycles is greater than or equal to "pulse_cycles" before proceed with the subtraction?

uLipe reviewed (commented):
LGTM so far, just consider the point from @wmrsouza  and I will also approve it.

mmahadevan108 commented:
@epc-ake Can you please add an issue with details and link that issue to this PR. After RC2 only bug fixes with an associated Github ssue will be merged into the upcoming 4.0 release.

raffarost commented:
hi @epc-ake,
can you check that you really are able to get a 100% duty with this solution?
I checked with the oscilloscope and see a pulse (99% duty, which is expected), so we think this might not be the best solution.
can you confirm?

epc-ake commented:
> hi @epc-ake, can you check that you really are able to get a 100% duty with this solution? I checked with the oscilloscope and see a pulse (99% duty, which is expected), so we think this might not be the best solution. can you confirm?
> 
> Please have a look at #80878 if possible

Thanks for pointing this out. I didn't notice it actually. 
I pushed a fix now but we can also use the changed from your PR. 
However, I have an jet an other issue with this driver. It seems like the driver ignores the output states defined in the pinctrl definition of the devicetree. e.g. `output-high;` has no effect. 
It's kind of important  for an inverted pwm to also initialize with 100% dut i think. 

@raffarost please tell me how you'd like to proceed with this and I'll adopt.

raffarost commented:
@epc-ake I think the way you implemented is also alright, but I'd personally leave the possibility to set the value in the DT, including the max res (for applications which won't use duty = 100%). Probably not that relevant but I think leaving open the possibility of using the config is better.

@sylvioalves what do you think?

epc-ake commented:
I suggest that I remove my changes regarding 100% pwm issue so we can use your PR for that. Looks cleaner to me. 
Meanwile I will try to fix the initialization state issue for inverted pwm polarity in this PR.

raffarost commented:
hi @epc-ake,
I spoke to Sylvio about this and we concluded you can keep the fix in your PR.
Would you mind adding a comment to explain why `SOC_LEDC_TIMER_BIT_WIDTH - 1` is used?
thank you

raffarost commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -312,7 +316,7 @@ static int pwm_led_esp32_set_cycles(const struct device *dev, uint32_t channel_i
 	double duty_cycle = (double) pulse_cycles / (double) period_cycles;
 
-	channel->duty_val = (uint32_t)((double) (1 << channel->resolution) * duty_cycle);
+	channel->duty_val = (uint32_t)((double)((1 << channel->resolution)) * duty_cycle);
```
can you revert to the original line?

epc-ake committed:
drivers: pwm_led: esp32: fix 100 percent pwm and init state

Set pwm init state to value defined in flags field.
Furthermore 100 percent is now actually 100 percent and not 99.

Signed-off-by: Armin Kessler <ake@espros.com>

epc-ake commented:
Hi all
I needed to refactor some code here to be able to initialize the pwm with high state (needed for inverted pwm's)
This includes moving the channel and pin configuration out of `pwm_led_esp32_set_cycles()` to `pwm_led_esp32_init()`. 
The only way I found setting the initial state of the pwm's was by using the `ledc_hal_set_idle_level()` function in combination with a new dt-property `init-high`.

I intentionally made this changes in a separate commit but if you prefer, I can also squeeze them into one. 
@raffarost and @sylvioalves, could you have a look?

raffarost commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -361,15 +369,21 @@ static const struct pwm_driver_api pwm_led_esp32_api = {
+		.speed_mode = DT_REG_ADDR(node_id) < SOC_LEDC_CHANNEL_NUM ? LEDC_LOW_SPEED_MODE    \
+									  : !LEDC_LOW_SPEED_MODE,  \
+		.clock_src = CLOCK_SOURCE,                                                         \
+		.inverted = DT_PROP(node_id, init_high),                                           \
```
I couldn't be sure this flag would fit in the field `flags` already foreseen in the DT (pwm_dt_spec / pwm_flags_t)
It would align better with the runtime check made in function `set_cycles`

epc-ake replied:
Do you know how to get the value of the flags field here?
I‘d actually prefer that over adding an extra dt-property but I didn’t find a way to do it.

raffarost replied:
in `include/zephyr/drivers/pwm.h` stay all the macros to unpack dts values:
```
 *    const struct pwm_dt_spec spec =
 *        PWM_DT_SPEC_GET_BY_NAME(DT_NODELABEL(n), alpha);
 *
 *    // Initializes 'spec' to:
 *    // {
 *    //         .dev = DEVICE_DT_GET(DT_NODELABEL(pwm1)),
 *    //         .channel = 1,
 *    //         .period = 1000,
 *    //         .flags = PWM_POLARITY_NORMAL,
 *    // }
 ```
 does it make sense to declare a pwms field for this driver (and include the flag)?
 I don't know enough about the API to give a solid opinion now, but others might know better.
 
 ```
 pwms = <&pwm2 3 2000 PWM_POLARITY_INVERTED>;
 ```

raffarost reviewed (commented):
LGTM
you'll need to rebase it with latest `main`

raffarost commented:
@epc-ake do you intend to keep working on this PR?
we shall at some point add these functionalities to the driver, so just asking to have an idea

epc-ake commented:
Hi @raffarost. 
I'm a bit busy currently but I might find some time at the end of this week for this. 
If you want to speed it up feel free to create a new PR on top of this :-)

raffarost commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -365,15 +368,21 @@ static DEVICE_API(pwm, pwm_led_esp32_api) = {
+		.speed_mode = DT_REG_ADDR(node_id) < SOC_LEDC_CHANNEL_NUM ? LEDC_LOW_SPEED_MODE    \
+									  : !LEDC_LOW_SPEED_MODE,  \
+		.clock_src = CLOCK_SOURCE,                                                         \
+		.inverted = DT_PWMS_FLAGS(node_id),                                                \
```
regarding this field, I checked other `PWM` drivers and these macros and `pwms` field seems to be only used by application, not directly by the driver. Considering this apparent 'standard' for the drivers, I'm not sure we can use it here. Possibly your earlier solution was more appropriate.

sylvioalves replied:
Sounds the earlier solution is indeed more appropriate.

raffarost commented:
> > @epc-ake I did some testing and think a fix regarding frequency update is needed (please check #82820, last commit). I tested on all devices and it seems to work fine. If you agree and can also kindly test them, feel free to incorporate the changes in your PR.
> 
> @raffarost Since you already made a PR on top of this one (#82820), would you take over this? I wont find enough time this year any more.

hi @epc-ake @braiden
sure, I'll evaluate the issues you mentioned and work on it.
thanks for the input
